packages/protocol/contracts/identity/IdentityProxyHub.sol

```diff
@@ -0,0 +1,99 @@
+pragma solidity ^0.5.13;
+
+import "../common/Create2.sol";
+import "../common/UsingRegistry.sol";
+import "./IdentityProxy.sol";
+
+contract IdentityProxyHub is UsingRegistry {
+  bytes32 public constant identityProxyCodeHash = keccak256(
+    abi.encodePacked(type(IdentityProxy).creationCode)
+  );
+
+  mapping(bytes32 => IdentityProxy) public identityProxies;
+
+  /**
+   * @notice Returns the IdentityProxy address corresponding to the identifier.
+   * @param identifier The identifier whose proxy address is computed.
+   * @return The identifier's IdentityProxy address.
+   * @dev This function will return a correct address whether or not the
+   * corresponding IdentityProxy has been deployed. IdentityProxies are deployed
+   * using CREATE2 and this computes a CREATE2 address.
+   */
+  function getIdentityProxy(bytes32 identifier) public view returns (IdentityProxy) {
+    return IdentityProxy(Create2.computeAddress(address(this), identifier, identityProxyCodeHash));
+  }
+
+  /**
+   * @notice Returns the IdentityProxy address corresponding to the identifier,
+   * deploying an IdentityProxy if one hasn't already been deployed for this
+   * identifier.
+   * @param identifier The identifier whose proxy address is returned.
+   * @return The identifier's IdentityProxy address.
+   */
+  function getOrDeployIdentityProxy(bytes32 identifier) public returns (IdentityProxy) {
+    IdentityProxy identityProxy = identityProxies[identifier];
+    if (address(identityProxy) == address(0)) {
+      identityProxy = deployIdentityProxy(identifier);
+      identityProxies[identifier] = identityProxy;
+    }
+
+    return identityProxy;
+  }
+
+  /**
+   * @notice Returns true if the given address is the likely owner of the given
+   * identifier.
+   * @param addr The address to check.
+   * @param identifier The identifier to check.
+   * @return True if the given address is the likely owner of the given
+   * identifier, false otherwise.
+   */
+  function passesIdentityHeuristic(address addr, bytes32 identifier) public returns (bool) {
+    IAttestations attestations = getAttestations();
+    (uint32 completed, uint32 requested) = attestations.getAttestationStats(identifier, addr);
+
+    bool hasEnoughCompletions = completed >= 3;
+
+    bool completedOverHalfRequests = false;
+    if (completed > 0) {
+      completedOverHalfRequests = requested / completed < 2;
+    }
+
+    bool hasMostCompletions = true;
+    address[] memory addresses = attestations.lookupAccountsForIdentifier(identifier);
+    for (uint256 i = 0; i < addresses.length; i++) {
+      address otherAddr = addresses[i];
+      if (otherAddr != addr) {
+        (uint32 otherCompleted, uint32 _requested) = attestations.getAttestationStats(
+          identifier,
+          otherAddr
+        );
+        if (otherCompleted > completed) {
+          hasMostCompletions = false;
+          break;
+        }
```

Comment:
Now for a totally unimportant suggestion
```suggestion
        hasMostCompletions = hasMostCompletions && otherCompleted <= completed
```

Reply:
don't we want the short-circuit from the `break`?

Reply:
In my opinion, no. We should optimize for the passing case, and I don't think we need to worry about gas usage in the failing case. The suggested change gets rid of a conditional (i.e. a `JUMPI`), replacing it with an `AND`, on every iteration to save a bit gas, and it makes the control flow a bit easier to reason about.